File: mobile/ios/HnnpBleBroadcaster.swift
import Foundation
import CoreBluetooth

/// HnnpBleBroadcaster coordinates token rotation and BLE advertising for HNNP v2.
///
/// Requirements from spec.md (v2):
/// - Rotate tokens every 15 seconds using time_slot.
/// - Broadcast current packet frequently in active mode (OS-level interval ~300–500 ms).
/// - In background mode, advertising can be slowed or disabled.
/// - BLE advertising interval is independent from token rotation.
///
/// This class uses a 1-second timer to track time_slot boundaries and updates
/// the advertising payload only when the slot changes. CoreBluetooth controls
/// the actual radio interval.
final class HnnpBleBroadcaster: NSObject, CBPeripheralManagerDelegate {

    enum Mode {
        case active
        case background
        case stopped
    }

    private let peripheralManager: CBPeripheralManager
    private var mode: Mode = .stopped
    private var timer: Timer?
    private var currentTimeSlot: Int64?

    override init() {
        self.peripheralManager = CBPeripheralManager(delegate: nil, queue: nil)
        super.init()
        self.peripheralManager.delegate = self
    }

    func startActive() {
        mode = .active
        startTimerIfNeeded()
    }

    func startBackground() {
        mode = .background
        startTimerIfNeeded()
    }

    func stop() {
        mode = .stopped
        stopAdvertising()
        stopTimer()
    }

    private func startTimerIfNeeded() {
        guard timer == nil else { return }
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.tick()
        }
    }

    private func stopTimer() {
        timer?.invalidate()
        timer = nil
    }

    private func tick() {
        guard mode != .stopped else { return }
        guard peripheralManager.state == .poweredOn else { return }

        let now = Date().timeIntervalSince1970
        let slot = TimeSlotHelper.compute(unixTime: now)

        let previous = currentTimeSlot
        if previous == nil || previous != slot {
            currentTimeSlot = slot

            #if DEBUG
            // Debug-only logging for time_slot transitions; no secrets logged.
            NSLog("HnnpBleBroadcaster: time_slot changed from \(previous.map(String.init) ?? \"none\") to \(slot) (unixTime=\(now))")
            #endif

            // Immediately recompute token + MAC and restart advertising for the new slot.
            updateAdvertising(for: slot)
        }
    }

    private func updateAdvertising(for timeSlot: Int64) {
        do {
            let deviceSecret = try DeviceSecretManager.getOrCreateDeviceSecret()
            let deviceAuthKey = DeviceSecretManager.deriveDeviceAuthKey(from: deviceSecret)

            let presenceToken = try TokenGenerator.derivePresenceToken(deviceAuthKey: deviceAuthKey, timeSlot: timeSlot)

            let version: UInt8 = 0x02
            let flags: UInt8 = 0x00
            let ts32 = UInt32(clamping: timeSlot)

            let mac = try BlePacketBuilder.computeMac(
                deviceAuthKey: deviceAuthKey,
                version: version,
                flags: flags,
                timeSlot: ts32,
                tokenPrefix: presenceToken.tokenPrefix
            )

            let packet = try BlePacketBuilder.buildBlePacketV2(
                version: version,
                flags: flags,
                timeSlot: ts32,
                tokenPrefix: presenceToken.tokenPrefix,
                mac: mac
            )

            let advData = BleAdvertisingPayload.buildServiceData(from: packet)

            // Restart advertising with updated payload for this time_slot.
            peripheralManager.stopAdvertising()

            switch mode {
            case .active:
                peripheralManager.startAdvertising(advData)
            case .background:
                // In background mode, platform may already throttle advertising.
                // Optionally, we could stop advertising entirely to reduce power.
                peripheralManager.startAdvertising(advData)
            case .stopped:
                break
            }

        } catch {
            // Log only high-level error; never secrets.
            NSLog("HnnpBleBroadcaster: failed to update advertising for slot \(timeSlot): \(error.localizedDescription)")
        }
    }

    private func stopAdvertising() {
        peripheralManager.stopAdvertising()
    }

    // MARK: - CBPeripheralManagerDelegate

    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        // When powered on, the next timer tick will begin advertising as needed.
        if peripheral.state != .poweredOn {
            stopAdvertising()
        }
    }
}

File: nearid/mobile-old/android/HnnpBleBroadcaster.kt
package com.hnnp.device

import android.bluetooth.BluetoothAdapter
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.util.Log
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * HnnpBleBroadcaster coordinates token rotation and BLE advertising.
 *
 * Requirements from spec.md (v2):
 * - Rotate tokens every 15 seconds using time_slot.
 * - Broadcast current packet every ~300–500 ms in active mode (via OS advertising mode).
 * - In background mode, advertising can be slowed (~1–2s) or disabled.
 * - BLE advertising interval is independent from token rotation.
 *
 * This class uses a 1-second scheduler to track time_slot boundaries and updates
 * the advertising payload only when the slot changes. The underlying BLE
 * advertising interval is controlled by Android via AdvertiseSettings.
 */
class HnnpBleBroadcaster(
    private val context: Context,
    private val bluetoothAdapter: BluetoothAdapter,
) {

    enum class Mode {
        ACTIVE,
        BACKGROUND,
        STOPPED,
    }

    private val advertiser: BluetoothLeAdvertiser? = bluetoothAdapter.bluetoothLeAdvertiser
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    @Volatile
    private var mode: Mode = Mode.STOPPED

    @Volatile
    private var currentTimeSlot: Long? = null

    private val advertiseCallback = object : AdvertiseCallback() {}

    private var schedulerStarted = false

    fun startActive() {
        mode = Mode.ACTIVE
        ensureScheduler()
    }

    fun startBackground() {
        mode = Mode.BACKGROUND
        ensureScheduler()
    }

    fun stop() {
        mode = Mode.STOPPED
        stopAdvertising()
    }

    private fun ensureScheduler() {
        if (!schedulerStarted) {
            schedulerStarted = true
            scheduler.scheduleAtFixedRate(
                { tick() },
                0L,
                1L,
                TimeUnit.SECONDS,
            )
        }
    }

    private fun tick() {
        if (mode == Mode.STOPPED) {
            return
        }

        val unixTimeSeconds = System.currentTimeMillis() / 1000.0
        val slot = TimeSlot.compute(unixTimeSeconds)

        val previous = currentTimeSlot
        if (previous == null || previous != slot) {
            currentTimeSlot = slot

            // Debug-only logging for time_slot transitions; no secrets logged.
            if (android.os.BuildConfig.DEBUG) {
                Log.d(
                    "HnnpBleBroadcaster",
                    "time_slot changed from ${previous ?: "none"} to $slot (unixTime=$unixTimeSeconds)"
                )
            }

            // Immediately recompute token + MAC and restart advertising for the new slot.
            updateAdvertisingForSlot(slot)
        }
    }

    private fun updateAdvertisingForSlot(timeSlot: Long) {
        val adv = advertiser ?: return

        // Build new packet for this time_slot.
        try {
            val deviceSecret = DeviceSecretManager.getOrCreateDeviceSecret(context)
            val deviceAuthKey = DeviceSecretManager.deriveDeviceAuthKey(deviceSecret)

            val presenceToken = TokenGenerator.derivePresenceToken(deviceAuthKey, timeSlot)

            val version: Byte = 0x02
            val flags: Byte = 0x00

            val mac = BlePacketBuilder.computeMac(
                deviceAuthKey = deviceAuthKey,
                version = version,
                flags = flags,
                timeSlot = timeSlot,
                tokenPrefix = presenceToken.tokenPrefix,
            )

            val packet = BlePacketV2(
                version = version,
                flags = flags,
                timeSlot = timeSlot,
                tokenPrefix = presenceToken.tokenPrefix,
                mac = mac,
            )

            val advertiseData = BleAdvertisingPayload.buildServiceData(packet)
            val settings = buildSettingsForMode(mode)

            // Restart advertising with the new payload for this time_slot.
            adv.stopAdvertising(advertiseCallback)
            adv.startAdvertising(settings, advertiseData, advertiseCallback)
        } catch (e: Exception) {
            // Only log high-level error, never secrets.
            Log.w("HnnpBleBroadcaster", "Failed to update advertising for slot $timeSlot: ${e.message}")
        }
    }

    private fun stopAdvertising() {
        val adv = advertiser ?: return
        try {
            adv.stopAdvertising(advertiseCallback)
        } catch (_: Exception) {
            // Ignore errors on stop.
        }
    }

    private fun buildSettingsForMode(mode: Mode): AdvertiseSettings {
        val builder = AdvertiseSettings.Builder()
            .setConnectable(false)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_LOW)

        when (mode) {
            Mode.ACTIVE -> {
                // Approx. 100–300 ms interval depending on device/OS; within the spec's
                // desired 300–500 ms range for active mode.
                builder.setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
            }
            Mode.BACKGROUND -> {
                // Slower advertising (~1–2s) suitable for background mode.
                builder.setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_POWER)
            }
            Mode.STOPPED -> {
                builder.setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_POWER)
            }
        }

        return builder.build()
    }
}
